spec/sanitized_file_spec.rb

```diff
@@ -59,322 +60,337 @@
     end
   end
 
-  describe '#basename' do
+  describe "#basename" do
     it "should return the basename for complicated extensions" do
-      @sanitized_file = CarrierWave::SanitizedFile.new(file_path('complex.filename.tar.gz'))
-      expect(@sanitized_file.basename).to eq("complex.filename")
+      sanitized_file = CarrierWave::SanitizedFile.new(file_path("complex.filename.tar.gz"))
+
+      expect(sanitized_file.basename).to eq("complex.filename")
     end
 
     it "should be the filename if the file has no extension" do
-      @sanitized_file = CarrierWave::SanitizedFile.new(file_path('complex'))
-      expect(@sanitized_file.basename).to eq("complex")
+      sanitized_file = CarrierWave::SanitizedFile.new(file_path("complex"))
+
+      expect(sanitized_file.basename).to eq("complex")
     end
   end
 
-  describe '#extension' do
+  describe "#extension" do
     %w[gz bz2 z lz xz].each do |ext|
       it "should return the extension for complicated extensions (tar.#{ext})" do
-        @sanitized_file = CarrierWave::SanitizedFile.new(file_path("complex.filename.tar.#{ext}"))
-        expect(@sanitized_file.extension).to eq("tar.#{ext}")
+        sanitized_file = CarrierWave::SanitizedFile.new(file_path("complex.filename.tar.#{ext}"))
+
+        expect(sanitized_file.extension).to eq("tar.#{ext}")
       end
     end
 
     it "should return the extension for real-world user file names" do
-      @sanitized_file = CarrierWave::SanitizedFile.new(file_path('Photo on 2009-12-01 at 11.12.jpg'))
-      expect(@sanitized_file.extension).to eq("jpg")
+      sanitized_file = CarrierWave::SanitizedFile.new(file_path("Photo on 2009-12-01 at 11.12.jpg"))
+
+      expect(sanitized_file.extension).to eq("jpg")
     end
 
     it "should return the extension for basic filenames" do
-      @sanitized_file = CarrierWave::SanitizedFile.new(file_path('something.png'))
-      expect(@sanitized_file.extension).to eq("png")
+      sanitized_file = CarrierWave::SanitizedFile.new(file_path("something.png"))
+
+      expect(sanitized_file.extension).to eq("png")
     end
 
     it "should be an empty string if the file has no extension" do
-      @sanitized_file = CarrierWave::SanitizedFile.new(file_path('complex'))
-      expect(@sanitized_file.extension).to eq("")
+      sanitized_file = CarrierWave::SanitizedFile.new(file_path("complex"))
+
+      expect(sanitized_file.extension).to eq("")
     end
   end
 
-  describe '#filename' do
-
-    before do
-      @sanitized_file = CarrierWave::SanitizedFile.new(nil)
-    end
+  describe "#filename" do
+    let(:sanitized_file) { CarrierWave::SanitizedFile.new(nil) }
 
     it "should default to the original filename if it is valid" do
-      expect(@sanitized_file).to receive(:original_filename).at_least(:once).and_return("llama.jpg")
-      expect(@sanitized_file.filename).to eq("llama.jpg")
+      expect(sanitized_file).to receive(:original_filename).at_least(:once).and_return("llama.jpg")
```

Comment:
We seem to rely on mocking a fair bit in this spec. Could you not instead use a `Tempfile`?

```ruby
describe "#filename" do
  let(:sanitized_file) { CarrierWave::SanitizedFile.new(file) }

  subject { sanitized_file.filename }

  context "when the original filename is valid"
    let(:file) { Tempfile(["image", ".jpg"]) }
    let(:filename) { file.path.split("/").last }

    specify { expect(subject).to eq filename }
  end

  context "when the original filename contains illegal characters"
    let(:file) { Tempfile(["a%b#c?d", ".jpg"]) }
    let(:filename) { file.path.split("/").last }

    specify { expect(subject).to eq filename.gsub("a%b#c?d", "a_b_c_d") }
  end
end